pkg/zcapld/create.go

```diff
@@ -0,0 +1,240 @@
+	}
+}
+
+// WithCapabilityChain specifies the capabilityChain on the proof for the Capability.
+func WithCapabilityChain(chain ...interface{}) CapabilityOption {
+	return func(o *CapabilityOptions) {
+		o.CapabilityChain = chain
+	}
+}
+
+// Signer signs the Capability.
+type Signer struct {
+	ariessigner.SignatureSuite
+	SuiteType          string
+	VerificationMethod string
+}
+
+// NewCapability constructs a new, signed Capability with the options provided.
+func NewCapability(signer *Signer, options ...CapabilityOption) (*Capability, error) {
+	if signer == nil {
+		return nil, errors.New("must provide a signer")
+	}
+
+	opts := &CapabilityOptions{
+		ID: fmt.Sprintf("urn:uuid:%s", uuid.New().String()),
+	}
+
+	for i := range options {
+		options[i](opts)
+	}
+
+	zcap := &Capability{
+		Context:          SecurityContextV2,
+		ID:               opts.ID,
+		Invoker:          opts.Invoker,
+		Controller:       opts.Controller,
+		Delegator:        opts.Delegator,
+		Parent:           opts.Parent,
+		AllowedAction:    opts.AllowedAction,
+		InvocationTarget: opts.InvocationTarget,
+	}
+
+	err := signZCAP(zcap, signer, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign zcap: %w", err)
+	}
+
+	return zcap, nil
+}
+
+func signZCAP(zcap *Capability, sig *Signer, options *CapabilityOptions) error {
+	raw, err := json.Marshal(zcap)
+	if err != nil {
+		return fmt.Errorf("failed to marshal zcap: %w", err)
+	}
+
+	nonce, err := nonce()
+	if err != nil {
+		return fmt.Errorf("failed to generate a nonce: %w", err)
+	}
+
+	now := time.Now()
+	ldSigner := ariessigner.New(sig)
+
+	signedDoc, err := ldSigner.Sign(
+		&ariessigner.Context{
+			SignatureType:           sig.SuiteType,
+			SignatureRepresentation: proof.SignatureJWS,
+			Created:                 &now,
+			Domain:                  options.Domain,
+			Nonce:                   nonce,
+			VerificationMethod:      sig.VerificationMethod,
+			Challenge:               options.Challenge,
+			Purpose:                 ProofPurpose,
+			CapabilityChain:         options.CapabilityChain,
+		},
+		raw,
+	)
+	if err != nil {
+		return fmt.Errorf("document signer failed to sign zcap: %w", err)
+	}
+
+	zcap.Proof, err = parseProof(signedDoc)
+	if err != nil {
+		return fmt.Errorf("failed to parse proof for zcap: %w", err)
+	}
+
+	return nil
+}
+
+func nonce() ([]byte, error) {
+	n := make([]byte, 64)
+
+	_, err := rand.Reader.Read(n)
+	if err != nil {
+		return nil, fmt.Errorf("nonce: failed to read from crypto/rand: %w", err)
+	}
+
+	return n, nil
+}
+
+func parseProof(signedZcap []byte) ([]verifiable.Proof, error) {
+	rawProof := &struct {
+		Proof json.RawMessage `json:"proof,omitempty"`
+	}{}
+
+	err := json.Unmarshal(signedZcap, rawProof)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal zcap to extract proof: %w", err)
+	}
+
+	var singleProof verifiable.Proof
+
+	err = json.Unmarshal(rawProof.Proof, &singleProof)
+	if err == nil {
+		return []verifiable.Proof{singleProof}, nil
```

Comment:
George, can you plz add a test for a single proof case?

Reply:
[Framework is always setting an array of proofs](https://github.com/hyperledger/aries-framework-go/blob/f1c7f7aa39eca010c0c14cdb8ea7e0287db52368/pkg/doc/signature/proof/utils.go#L62-L81). I guess I will remove the "single proof" code path since `parseProof` is an internal function that cannot be accessed from `zcapld_test`.

Note: I just copied the framework's [logic](https://github.com/hyperledger/aries-framework-go/blob/65d4573d45857f82bd271b019e5c1eac8549e4ad/pkg/doc/verifiable/common.go#L265-L285).

Reply:
Done.